---
title: "Manipulating and plotting genomic data"
output:
  html_notebook:
    toc: true
    toc_float: true
---

We are going to use some nice packages to query, manipulate, and plot genomic data. The point of this case is to showcase the versatility of R and the breadth of application specific packages available. It is not the point that you should fully understand how these packages work, nor should you understand the genetics aspect.

Below, we install some packages, and note that we do not use the normal `install.packages()` command, but rather a command from `BiocInstaller`, which is itself a package that needs to be installed. All these packages are from the Bioconductor (https://www.bioconductor.org/) ecosystem of packages, which is an organization and a community that curates high quality R packages for reproducible research in genetics.

We install `GenomicRanges`, which is a data structure with powerful and convenient manipulation tools. All data in `GenomicRanges` objects is based on *intervals* of the genome, and we can very easily query about a specific region or find overlap between objects and much more.

The `TxDb.Hsapiens.UCSC.hg38.knownGene` package contains data, in the `GenomicRanges` format, with information about genes in the human genome.

`Gviz` gives us tools to easily plot genomic information.

```{r eval=FALSE}
BiocInstaller::biocLite('GenomicRanges')
BiocInstaller::biocLite('TxDb.Hsapiens.UCSC.hg38.knownGene')
BiocInstaller::biocLite('Gviz')
```

Load some libraries.

```{r}
library(magrittr)  # The pipe operator %>%.
library(readr)  # Reading data.
library(dplyr)  # Tibble manipulation.
library(TxDb.Hsapiens.UCSC.hg38.knownGene)  # Info about genes in humans.
library(GenomicRanges)  # Genomic data structure.
library(Gviz)  # Vizualize genomic data.
```

# Query *TxDb* about human genome metadata

First, we want to get some data relating to a specific gene, called SCN5A, which is implicated in irritable bowel disease. The database contains, among other things, information about *transcripts* and *exons*, and what we want is all the exons corresponding to a specific transcript of SCN5A.

So below we basically query the database for transcripts of gene SCN5A, then we pick a transcript, query the database for all exons, and extract the exons corresponding to the transcript of interest.

```{r}
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene

# UCSC/Entrez ID of SCN5A.
gene_id <- '6331'

# Get transcripts corresponding to the gene.
tx <- transcriptsBy(txdb, by='gene')  # All transcripts by gene.
tx <- tx[[gene_id]]

# Get the exons corresponding to one of the transcripts in the gene.
exons <- exonsBy(txdb, by='tx')  # All exons by transcript.
tx_name <- tx$tx_name[1]  # Pick the first transcript.
tx_id <- tx$tx_id[1]
exons <- exons[[tx_id]]
```

Let's see what this `GenomicRanges` data object looks like. Note that it looks a lot like a dataframe or a tibble, with columns, rows and data classes. Note also that is has some extra information about the data in the table itself.

```{r}
head(exons)
```

# Simple *Gviz* plots

Let's make a simple plot with `Gviz`. Below, we make an "annotation track", supplying the exons and a track name, and tell `Gviz` to plot this track. Quite boring plot, let's make it a little bit more interesting.

```{r fig.width=5, fig.height=2}
# Make a genomic track.
atrack <- AnnotationTrack(exons, name="Exons of SCN5A")
# Plot track.
plotTracks(atrack)
```

There are many different types of tracks in `Gviz`, and below we examples of an ideogram, which shows our location in the genome, a genome axis track, showing more detailed positions, and a gene region track, again showing the exons but a little bit differently than before. We can pass a lot of different arguments to these tracks and the `plotTracks()` function to modify our plots.

```{r fig.width=8, fig.height=4}
# Get chromosome and genome (species and version) from exons object.
chr <- as.character(unique(seqnames(exons)))
gen <- as.character(unique(genome(exons)))
# Ideogram track shows where in the chromosome we are located.
itrack <- IdeogramTrack(genome=gen, chromosome=chr)
# Axis track shows position on chromosome.
gtrack <- GenomeAxisTrack()
# Gene region track shows general inforation about intervals on the genome. Use it to plot exons.
grtrack <- GeneRegionTrack(exons, genome=gen, chromosome=chr, name='Exons of SCN5A')
# Plot all tracks, and specify relative sizes (proportions) of tracks.
plotTracks(list(itrack, gtrack, grtrack), sizes=c(0.2,0.2,0.6))
```

# Experimental data and more about *GenomicRanges*

We are going to download some data to plot in this gene. The GWAS Catalog (https://www.ebi.ac.uk/gwas/) contains summary results from research in the genetic causes of disease, aggregated from many independent experiments. This data contains the location of mutations in the genome and their reported effect size from regression analysis, indicating increase in disease risk.

Below we read in a table with these data, using `read_delim()` which will store the data in a tibble, as we have seen in the previous cases. This dataset is quite large, so it will take a while to read and use *a lot* of memory.

```{r eval=FALSE}
# Read experimental data from ulcerative colitis.
uc = read_delim('~/Documents/gwas_sumstats/uc/28067908-GCST004133-EFO_0000729.h.tsv.gz', delim='\t')
```

We happen to know that the gene we are interested in lies on chromosome 3, so we can simply discard everything else, so we can free up a lot of memory on the computer. We do this using the `filter()` function, which chooses rows in the tibble that match the expression we pass to it. In this expression, we use the `hm_chrom` variable, which tells us which chromosome number the mutations is located on.

```{r eval=FALSE}
# Keep records that match expression. Choose records in chromosome 3.
uc <- uc %>% filter(hm_chrom == 3)
```

When working with large datasets, `rm()` and `gc()` are your friend. `rm()` removes the variable you pass to it, and `gc()` is "garbage collection" and frees up unused memory. Since we just removed a bunch of rows from `uc`, we are going to use `gc()` to clear up that memory.

```{r eval=FALSE}
# Free up unused memory.
gc()
```

We do the same for the Crohn's disease data.

```{r eval=FALSE}
# Read experimental data from Crohn's disease.
cd = read_delim('~/Documents/gwas_sumstats/cd/28067908-GCST004132-EFO_0000384.h.tsv.gz', delim='\t')
# Choose records on chromosome 3.
cd <- cd %>% filter(hm_chrom == 3)
# Free up unused memory.
gc()
```

These datasets contain many columns that we are not interested in. Let's delete all these, and give the variables more convenient names.

```{r eval=FALSE}
# Remove irrelevant columns.
# Give columns more convenient names.
uc = uc %>% dplyr::select(c('hm_rsid', 'hm_chrom', 'hm_pos', 'hm_beta')) %>% rename('hm_rsid'='rsid', 'hm_chrom'='chrom', 'hm_pos'='pos', 'hm_beta'='beta')
cd = cd %>% dplyr::select(c('hm_rsid', 'hm_chrom', 'hm_pos', 'hm_beta')) %>% rename('hm_rsid'='rsid', 'hm_chrom'='chrom', 'hm_pos'='pos', 'hm_beta'='beta')
```

The remaining variables are the IDs of the mutations, the "rsid", the chromosome number where the mutation is located, the position along this chromosome, and the effect size beta.

We want to have all this data in the same tibble, and we do this using a *join* operation. `full_join()` takes two datasets and matches by the variables you supply; the rest of the variables that weren't mentioned in the `by` argument are added to the resulting dataset with unique names. Anything that doesn't match anything else will make NA values in your data. We also give the betas intuitive names below.

```{r eval=FALSE}
# Joint two tibbles by: rsid, chrom and pos.
# full_join: "by" variables with no matches get NA values in unmatched dataset.
betas = full_join(cd, uc, by=c('rsid'='rsid', 'chrom'='chrom', 'pos'='pos'))
# Give more convenient names to betas.
betas <- betas %>% rename('beta.x'='beta_cd', 'beta.y'='beta_uc')
```

Now we're done with `uc` and `cd`, since we have all the information in `betas`, we delete the variables and free up the memory.

```{r eval=FALSE}
# Delete variables and free up memory.
rm(uc, cd)
gc()
```

The chromosomes in the data are just integers, but want them to be in the format "chr3", so we use the `paste0()` function to do this job for us, and *mutate* the variable in the dataset. `paste0()` takes basically any values or vectors and concatenates them into a character string, this function is very often useful.

```{r eval=FALSE}
# Use "chr3" naming convention, same as used in the exons object.
betas <- betas %>% mutate(chrom=paste0('chr', chrom))
```

We want to turn this tibble into the `GenomicRanges` data structure, so that we can work with it together with the exon data. We construct this using the `GRanges()` function, supplying data from the tibble about the locations of the mutations. Then we add the extra information about the mutations, the metadata, namely the effect sizes in Crohn's disease and ulcerative colitis.

```{r eval=FALSE}
# Make a GRanges object out of the betas.
betas_gr = GRanges(seqnames=betas$chrom, IRanges(start=betas$pos, end=betas$pos))
# Add the metadata.
mcols(betas_gr, use.names=TRUE) = dplyr::select(betas, 'beta_cd', 'beta_uc')
```

Next we want to remove all betas that are not inside the gene region we are interested in. This means we have to construct a new `GenomicRanges` object, that spans the entire gene region. When we've done that we use `subsetByOverlaps()` to retrieve the betas that are in the gene region.

```{r eval=FALSE}
# Extract betas in the SCN5A gene.
# Get the first and last positions in the exons.
start <- min(start(exons))
end <- max(end(exons))
# Get the chromosome.
chr <- as.character(unique(seqnames(exons)))
gene_region <- GRanges(seqnames=chr, ranges=IRanges(start=start, end=end))
# Get the subset of the betas that are in the gene region.
betas_gr <- subsetByOverlaps(betas_gr, gene_region)
```

Now we again delete some data.

```{r eval=FALSE}
# Delete variables and free up memory.
rm(betas)
gc()
```

All the code dealing with the CD and UC data takes a long time to run and uses a lot of memory, so we want to save our results so far. We convert the `GenomicRanges` object to a tibble and write a CSV with the data.

```{r eval=FALSE}
# Convert GRanges object to tibble and write to CSV.
write_csv(as_tibble(betas_gr), 'betas.csv')
```

To load the data, we need to read it into a tibble, and construct the `GenomicRanges` object in a similar way as we did before.

```{r}
# Read CSV into tibble.
betas_temp <- read_csv('betas.csv')
# Construct GRanges object and add betas.
betas_gr = GRanges(seqnames=betas_temp$seqnames, IRanges(start=betas_temp$start, end=betas_temp$end))
mcols(betas_gr, use.names=TRUE) = dplyr::select(betas_temp, 'beta_cd', 'beta_uc')
```

# Plot experimental data

Finally, we plot everything. As before, we plot the ideogram, the axis, and the exons. Now we add a data track, pass it the `GenomicRanges` with the betas, and it will plot the beta values. Using the `groups=c('cd', 'uc')` argument, `Gviz` will match the columns with these strings and color the data by group.

```{r}
# Get chromosome and genome build from exons object.
chr <- as.character(unique(seqnames(exons)))
gen <- genome(exons)
# Ideogram track shows where in the chromosome we are located.
itrack <- IdeogramTrack(genome=gen[1], chromosome=chr)
# Axis track shows position on chromosome.
gtrack <- GenomeAxisTrack()
# Plot exons as gene regions.
grtrack <- GeneRegionTrack(exons, genome=gen, chromosome=chr, name='Exons of SCN5A')
dtrack <- DataTrack(betas_gr, groups=c('cd', 'uc'))
plotTracks(list(itrack, gtrack, grtrack, dtrack))
```

We did it! We plotted some experimental data along with genomic annotations! So what did we learn in this case?

* You can find great tools in R almost no matter what field you work in
* Packages like the ones in Bioconductor allow you do not only be more productive but let you create more reproducible research
* Packages in R sometimes contain useful metadata, in this case we used information about genes in the human genome
* Be careful when working with large datasets, and use `rm()` and `gc()` as much as you can
* We learned some about some new R functions:
  * The *filter* operation: `filter()`
  * The *join* operation: `full_join()`
  * The `paste0()` function

```{r}
devtools::session_info()
```



